src/smartpeak/include/SmartPeak/core/EventDispatcher.h

```diff
@@ -0,0 +1,118 @@
+// --------------------------------------------------------------------------
+//   SmartPeak -- Fast and Accurate CE-, GC- and LC-MS(/MS) Data Processing
+// --------------------------------------------------------------------------
+// Copyright The SmartPeak Team -- Novo Nordisk Foundation 
+// Center for Biosustainability, Technical University of Denmark 2018-2021.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL ANY OF THE AUTHORS OR THE CONTRIBUTING
+// INSTITUTIONS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+// OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+// OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+// ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// --------------------------------------------------------------------------
+// $Maintainer: Bertrand Boudaud $
+// $Authors: Douglas McCloskey, Bertrand Boudaud $
+// --------------------------------------------------------------------------
+
+#pragma once
+
+#include <SmartPeak/iface/IApplicationProcessorObserver.h>
+#include <SmartPeak/core/ApplicationProcessorObservable.h>
+#include <SmartPeak/iface/ISequenceProcessorObserver.h>
+#include <SmartPeak/core/SequenceProcessorObservable.h>
+#include <SmartPeak/iface/ISampleGroupProcessorObserver.h>
+#include <SmartPeak/core/SampleGroupProcessorObservable.h>
+#include <SmartPeak/iface/ISequenceSegmentProcessorObserver.h>
+#include <SmartPeak/core/SequenceSegmentProcessorObservable.h>
+#include <SmartPeak/iface/ISequenceObserver.h>
+#include <SmartPeak/core/SequenceObservable.h>
+#include <SmartPeak/iface/ITransitionsObserver.h>
+#include <SmartPeak/core/TransitionsObservable.h>
+
+#include <plog/Log.h>
+#include <future>
+#include <thread>
+#include <string>
+#include <vector>
+
+namespace SmartPeak
+{
+
+  /**
+    The EventDispatcher can be used to store and dispatch Observer events to another thread.
+  */
+  struct EventDispatcher :
+    IApplicationProcessorObserver,
+    ISequenceProcessorObserver,
+    ISequenceSegmentProcessorObserver,
+    ISampleGroupProcessorObserver,
+    ISequenceObserver,
+    ITransitionsObserver,
+    ApplicationProcessorObservable,
+    SequenceProcessorObservable,
+    SequenceSegmentProcessorObservable,
+    SampleGroupProcessorObservable,
+    SequenceObservable,
+    TransitionsObservable
+  {
+    EventDispatcher() = default;
+
+    /**
+      IApplicationProcessorObserver
+    */
+    virtual void onApplicationProcessorStart(const std::vector<std::string>& commands) override;
```

Comment:
I don't see the need to split into different classes, do you have specific in mind (a concern?). having a unique dispatcher looks correct, and also in this way the event are stored and dispatched in order.

Reply:
From my reading of the code, it seemed that several of the methods only correspond to specific processors, so it might make sense to make processor specific classes that inherit common methods from a base class.